src/test/java/com/amihaiemil/zold/mock/MockHttpClient.java

```diff
@@ -0,0 +1,161 @@
+
+    /**
+     * Ctor.
+     *
+     * @param assertions Given assertions.
+     */
+    @SuppressWarnings("unchecked")
+    public MockHttpClient(final AssertRequest... assertions) {
+        Arrays.stream(assertions).forEach((assertion) -> {
+            this.assertions.add(assertion);
+        });
+    }
+
+    @Override
+    public HttpParams getParams() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    public ClientConnectionManager getConnectionManager() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    public HttpResponse execute(final HttpUriRequest request)
+        throws IOException, ClientProtocolException {
+        return this.check(request);
+    }
+
+    @Override
+    public HttpResponse execute(final HttpUriRequest request,
+        final HttpContext context)
+        throws IOException, ClientProtocolException {
+        return this.check(request);
+    }
+
+    @Override
+    public HttpResponse execute(final HttpHost target,
+        final HttpRequest request)
+        throws IOException, ClientProtocolException {
+        return this.check(request);
+    }
+
+    @Override
+    public HttpResponse execute(final HttpHost target,
+        final HttpRequest request, final HttpContext context)
+        throws IOException, ClientProtocolException {
+        return this.check(request);
+    }
+
+    @Override
+    public <T> T execute(final HttpUriRequest request,
+        final ResponseHandler<? extends T> responseHandler)
+        throws IOException, ClientProtocolException {
+        return responseHandler.handleResponse(
+            this.check(request)
+        );
+    }
+
+    @Override
+    public <T> T execute(final HttpUriRequest request,
+        final ResponseHandler<? extends T> responseHandler,
+        final HttpContext context)
+        throws IOException, ClientProtocolException {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    public <T> T execute(final HttpHost target, final HttpRequest request,
+        final ResponseHandler<? extends T> responseHandler)
+        throws IOException, ClientProtocolException {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    //@checkstyle ParameterNumber (8 lines)
+    @Override
+    public <T> T execute(final HttpHost target, final HttpRequest request,
+        final ResponseHandler<? extends T> responseHandler,
+        final HttpContext context)
+        throws IOException, ClientProtocolException {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    /**
+     * Checks all conditions against the request.
+     *
+     * @param request The request.
+     * @return HttpResponse if all the cehcks pass.
+     */
+    private HttpResponse check(final HttpRequest request) {
+        final Collection<Condition> conditions = this.assertions.get(0)
```

Comment:
@denis-konovalyenko An ``AssertRequest`` has some conditions and a response which is returned when the actual HttpRequest, that is performed against this mock client, respects the conditions. 

We only use the first ``AssertRequest`` and then remove it from the list since only one actual ``HttpRequest`` enters the HttpClient at a given time. 

Makes sense? :D (I know it might be a little twisted)

Reply:
@denis-konovalyenko And it is named ``AssertRequest`` so it makes more sense when actually using it in the tests. In the tests, when we instantiate an object, it will be like this:

```java
new RestfulZoldWts(
    new MockHttpClient (
        new AssertRequest(..., ...), //There are some conditions here
        new AssertRequest(...., ...) 
    ),
   ...
)
```
So you read the above as follows: "we assert that the first 2 HttpRequests performed by ``RestfulZoldWts`` respect the given conditions"

Reply:
@amihaiemil , I think it should be either got at once by reading the code or just avoided in order not to raise the level of complexity for almost no reason.

Thus, if a `MockHttpClient` is going to assert some requests for a particular resource, then they all have to be checked (tested), haven't they? If I am reading the code right, only the first `AssertRequest` will be checked. So, it would be better to either do not create the list of `AssertRequest`s, but have just one `AssertRequest`, or traverse all `AssertRequests` in the list.

By the way, how come that we are removing the first assertion (`AssertRequest`) from the list after testing its conditions? What wonders me most is that I can't take in what this is for?

Reply:
@denis-konovalyenko Each ``AssertRequest`` from that list coresponds to one actual ``HttpRequest`` that enters the HttpClient -- it is a 1:1 relationshio (and an ``AssertRequest`` encapsulates the conditions that have to be met by the ``HttpRequest`` that is tested). 

So, we only care about the first ``AssertRequest`` in the list at any given time, since it is testing the current HttpRequest. Then, we remove it from the list, since the next request that comes in has to be tested by the following ``AssertRequest``. 

Makes more sense? :D

Reply:
@amihaiemil , eventually, I have understood the idea behind these changes. We are going to pack a `MockHttpClient` with `AssertRequest`s and then gradually execute as many `HttpUriRequest`s as needed. However, I assume, we have to bear in mind that we can still get into the situation when there would be no more `AssertRequest`s in the list and we will end up with the `IndexOutOfBoundsException` on trying to get one from there. And this is not the case I would look for as an end user of the `MockHttpClient`. By the way, are you going to make the ability to execute more than 1 `HttpUriRequest` like this?
```
            new MockHttpClient(
                new AssertRequest(
                   ...
                ),
                new AssertRequest(
                   ...
                ),
            ).execute(
                new HttpGet(...),
                new HttpGet(...)
            ),
```

Could you please shed some light on why the implemented approach is better than performing the same with two separate `MockHttpClient`s:
```
            new MockHttpClient(
                new AssertRequest(
                   ...
                ),
            ).execute(
                new HttpGet(...)
            ),
...
            new MockHttpClient(
                new AssertRequest(
                   ...
                ),
            ).execute(
                new HttpGet(...)
            ),
```

Reply:
@denis-konovalyenko 
> However, I assume, we have to bear in mind that we can still get into the situation when there would be no more AssertRequests in the list and we will end up with the IndexOutOfBoundsException

Yes, this is right, if the object under test makes more HttpRequests than there are ``AssertRequests`` specified, there will be an exception -- I'll think about how we can make this a bit prettier. But, in principle, the person who writes the unit test should know how many requests the object will make.

> By the way, are you going to make the ability to execute more than 1 HttpUriRequest like this?

No, we don't need that.

> Could you please shed some light on why the implemented approach is better than performing the same with two separate MockHttpClients:

We only need one ``HttpClient``, always. And it is not us (in the tests) who calls ``MockHttpClient.execute(...)``, it is the object under test who calls it as if it would calling the real ``HttpClient``. So, the object sends requests to its encapsulated ``HttpClient`` (which we injected as ``MockHttpClient``) and then we run the conditions from each request's coresponding ``AssertRequest``. 

Well, the mere fact that it's so hard to understand seems like a bug in this testing design, but I have no idea how to do it otherwise and to me, it seems quite elegant. :D

Reply:
@amihaiemil , thanks for the quick response! All right, I understand the needs. Let's wrap up with that review, even the PR has been already merged. :)

Thank you for your contribution!